### PR TITLE
SOLR-16427: Evaluate and fix errorprone rules - part 4

### PR DESCRIPTION
**Pull request by risdenk**
https://issues.apache.org/jira/browse/SOLR-16427

builds on https://github.com/apache/solr/pull/1116

Enables the following errorprone rules:
* `EqualsUnsafeCast` - fixes 2 cases of not checking the type before casting
* `IdentityHashMapUsage` - fixes a few usages of `IdentityHashMap` to be the explicit type instead of `Map`
* `BadShiftAmount` - addressed a few shifts to not be negative
* `FloatingPointLiteralPrecision` - followed the error messages to fix floating point precision
* `CatchAndPrintStackTrace` - replaced with log instead of print stack trace
* `IntLongMath` - added casts to ensure no overflow
* `ImmutableEnumChecker` - made a few fields final and added suppressions for the rest
* `ObjectsHashCodePrimitive` - fixed a few boxing issues in hashcode
* `OperatorPrecedence` - add parenthesis to clarify boolean statements
* `InconsistentCapitalization` - address a few variable names to have consistent capitalization
* `IdentityBinaryExpression` - fix few cases with wrong comparison being made or not a number checks

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/api/collections/ReindexCollectionCmd.java [hunk 1]
```diff
@@ -138,7 +138,7 @@ public static State get(Object p) {
       return states.get(p);
     }
 
-    static Map<String, State> states =
+    static final Map<String, State> states =
         Collections.unmodifiableMap(
             Stream.of(State.values())
                 .collect(Collectors.toMap(State::toLower, Function.identity())));
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/ReindexCollectionCmd.java:144]: `Collectors.toUnmodifiableMap` will simplify this a little as well.

File: solr/core/src/java/org/apache/solr/filestore/DistribPackageStore.java [hunk 2]
```diff
@@ -72,16 +72,16 @@ public class DistribPackageStore implements PackageStore {
   private final CoreContainer coreContainer;
   private Map<String, FileInfo> tmpFiles = new ConcurrentHashMap<>();
 
-  private final Path solrhome;
+  private final Path solrHome;
 
   public DistribPackageStore(CoreContainer coreContainer) {
     this.coreContainer = coreContainer;
-    this.solrhome = Paths.get(this.coreContainer.getSolrHome());
+    this.solrHome = Paths.get(this.coreContainer.getSolrHome());
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/filestore/DistribPackageStore.java:79]: @sonatype-lift ignore

File: solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java [hunk 3]
```diff
@@ -1628,9 +1629,9 @@ public DirectoryFileStream(SolrParams solrParams) {
     }
 
     // Throw exception on directory traversal attempts
-    protected String validateFilenameOrError(String filename) {
-      if (filename != null) {
-        Path filePath = Paths.get(filename);
+    protected String validateFilenameOrError(String fileName) {
+      if (fileName != null) {
+        Path filePath = Paths.get(fileName);
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java:1634]: @sonatype-lift ignore

File: solr/core/src/java/org/apache/solr/cloud/api/collections/ReindexCollectionCmd.java (hunk 1 above)
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/ReindexCollectionCmd.java]: addressed in c324ae6e07fb6134708d9e67d38172ec91c4f27d

File: solr/core/src/java/org/apache/solr/schema/OpenExchangeRatesOrgProvider.java [hunk 4]
```diff
@@ -151,7 +151,7 @@ public boolean reload() throws SolrException {
     try {
       log.debug("Reloading exchange rates from {}", ratesFileLocation);
       try {
-        ratesJsonStream = (new URL(ratesFileLocation)).openStream();
+        ratesJsonStream = new URL(ratesFileLocation).openStream();
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/schema/OpenExchangeRatesOrgProvider.java:154]: @sonatype-lift ignoreall

**Review by risdenk — COMMENTED** [review]
Need to address a few things found on finer review

File: solr/core/src/test/org/apache/solr/cloud/TestTlogReplica.java [hunk 5]
```diff
@@ -679,7 +679,7 @@ public void testRecovery() throws Exception {
             waitingForReplay.release();
             waitingForBufferUpdates.acquire();
           } catch (InterruptedException e) {
-            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/test/org/apache/solr/cloud/TestTlogReplica.java:682]: Fixed in b5ccf7820968b3c24ad25e2ea56114b1d151ef1d

File: solr/core/src/test/org/apache/solr/cloud/TrollingIndexReaderFactory.java [hunk 6]
```diff
@@ -204,7 +203,7 @@ private static void recordStackTrace(StackTraceElement[] stackTrace) {
       try {
         lastStacktraces.poll(100, TimeUnit.MILLISECONDS);
       } catch (InterruptedException e1) {
-        e1.printStackTrace();
+        Thread.currentThread().interrupt();
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/test/org/apache/solr/cloud/TrollingIndexReaderFactory.java:206]: Fixed in b5ccf7820968b3c24ad25e2ea56114b1d151ef1d

File: solr/core/src/test/org/apache/solr/cloud/ZkShardTermsTest.java [hunk 7]
```diff
@@ -223,7 +223,7 @@ public void testRaceConditionOnUpdates() throws InterruptedException {
                       Thread.sleep(random().nextInt(200));
                       zkShardTerms.setTermEqualsToLeader(replica);
                     } catch (InterruptedException e) {
-                      e.printStackTrace();
+                      Thread.currentThread().interrupt();
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/test/org/apache/solr/cloud/ZkShardTermsTest.java:226]: Fixed in b5ccf7820968b3c24ad25e2ea56114b1d151ef1d

File: solr/core/src/test/org/apache/solr/response/TestRetrieveFieldsOptimizer.java [hunk 8]
```diff
@@ -411,21 +411,20 @@ static enum TEST_TYPE {
...
-                  new SimpleEntry<>(TEST_TYPE.TDATE, "solr.TrieDateField"),
-                  new SimpleEntry<>(TEST_TYPE.PINT, "solr.IntPointField"),
-                  new SimpleEntry<>(TEST_TYPE.PLONG, "solr.LongPointField"),
-                  new SimpleEntry<>(TEST_TYPE.PFLOAT, "solr.FloatPointField"),
-                  new SimpleEntry<>(TEST_TYPE.PDOUBLE, "solr.DoublePointField"),
-                  new SimpleEntry<>(TEST_TYPE.PDATE, "solr.DatePointField"),
-                  new SimpleEntry<>(TEST_TYPE.STRING, "solr.StrField"),
-                  new SimpleEntry<>(TEST_TYPE.BOOL, "solr.BoolField"))
-              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+      Stream.of(
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/response/TestRetrieveFieldsOptimizer.java:414]: No point in using Stream without any transformation.  Here, just call `Map.ofEntries`

File: solr/core/src/java/org/apache/solr/search/stats/LRUStatsCache.java [hunk 9]
```diff
@@ -129,7 +129,7 @@ protected ShardRequest doRetrieveStatsRequest(ResponseBuilder rb) {
     // force-fetched on next request and cached.
 
     // check for missing stats from previous requests
-    if (!missingColStats.isEmpty() || !missingColStats.isEmpty()) {
+    if (!missingColStats.isEmpty() || !missingTermStats.isEmpty()) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/stats/LRUStatsCache.java:132]: cool; perhaps an actual bug

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java [hunk 10]
```diff
@@ -87,7 +87,7 @@ public boolean isTripped() {
   @Override
   public String getDebugInfo() {
 
-    if (seenCPUUsage.get() == 0.0 || seenCPUUsage.get() == 0.0) {
+    if (seenCPUUsage.get() == 0.0 || allowedCPUUsage.get() == 0.0) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java:90]: cool; perhaps an actual bug

**Review by dsmiley — COMMENTED** [review]
Just a few things.

Thanks for doing this!

File: solr/core/src/test/org/apache/solr/response/TestRetrieveFieldsOptimizer.java (hunk 8 above)
Comment (risdenk) [reply solr/core/src/test/org/apache/solr/response/TestRetrieveFieldsOptimizer.java]: good catch - fixed in eb8e6ac281439f31ac3a59e247da5faccd7224dd

File: solr/core/src/java/org/apache/solr/parser/QueryParserTokenManager.java [hunk 11]
```diff
@@ -16,7 +17,9 @@
 
 
 
+
 /** Token Manager. */
+@Generated("JavaCC")
```
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/parser/QueryParserTokenManager.java:22]: fixed in 27ac23aa5df5f0a9e254f09ce12432d4a2c141da

**Comment by risdenk** [comment]
one last rebase to latest main to keep all the commit messages. no code changes in the PR. running tests again before merging.

**Review by epugh — APPROVED** [review]
It all looked good to me, I like the consistency and the additional verbosity to understand what is happening.     I couldn't evaluate the changes in _SOLR-16427: Evaluate and fix errorprone rules - BadShiftAmount_, as I didn't udnerstand how you got those magic numbers ;-)   I also don't love that the way we look up our log is via `private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());`  That is more of general comment that it always seems messy how we do that in Java in general, and isn't a request to change it.   *Logging*, sometimes should be referred to as *Clogging*  ;-).

**Comment by risdenk** [comment]
> SOLR-16427: Evaluate and fix errorprone rules - BadShiftAmount, as I didn't udnerstand how you got those magic numbers

well I didn't do the calculations :) errorprone did. The idea is that there is wraparound so the shifts can only be within a certain range. See https://errorprone.info/bugpattern/BadShiftAmount for an explanation.

> I also don't love that the way we look up our log is via private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

so historically logging in Solr was even more of a mess since the logger would point to the wrong class with copy/paste. This pattern at least ensures that the logger in one class will apply to that class - even if you copy paste to new code. So yea its a bit wonky but I'm a fan.

**Pull request merged**